### ISSUE TITLE
Problem creating the graph

### ISSUE THREAD

**Issue by frapik99** [axt/angr-utils#10]
Hi @axt,
I'm still trying to fix the problem with the ```angr-utils``` and i have a weird behaviour. Using ```pngpixel```,like last time, 'm trying to print a simple graph for ```png_user_version_check```, maybe I'm doing something wrong, but I cannot see the connection between the block ending in ```0x5a8``` and ```0x51d```. When you have time can you generate the graph for that function and upload here the result please?

Thanks

**Comment by frapik99** [comment]
Solved thanks

**Comment by frapik99** [comment]
HI @axt ,
I think I'm facing a problem but I cannot understand if it is real or not and if so if it is in angr or your project. Could you create the graph for ```png_user_version_check``` and tell me if you have an improper loop. I can see it when the ```png_get_header_ver``` is called the next block should be ```0x...589``` instead I have ```0x...584```. ```0x...584``` was the first block met when the ```png_get_header_ver``` was called the first time.

Thanks

**Comment by frapik99** [comment]
Hi @axt,
I have noticed if for example there are 2 calls to the ```puts``` function, if I generate the ```CFGFast``` the first block after the first call to the ```puts``` point to the first block after the second ```puts```. is it happening because I'm using ```CGFast```? I know that ```CFGFast``` guarantees the full coverage so I think this behaviour is incorrect. You can try to create the graph for the main ```function``` with the pngpixel I have shared with you last time.

Am I missing something?

Update: I think it might be an issue related to how angr creates the ```CFGFast```, I have raised an issue on github. IMO you could improving your project returning only the "real" CFG graph.

Thanks

**Comment by axt** [comment]
Hi @frapik99! Sorry, I dont have my dev-environment setup on this machine, I hopefully can check your first question on the weekend. 

About the second question: I'm not sure I understand what you mean:
CFGFast is unaware of calling context, so this behavior is normal. 
If you want a context-aware CFG try CFGAccurate instead, with context_sensitivity_level set to at least 1.

**Comment by axt** [comment]
I've tried to check what you asked for, but the version of pngpixel you've sent me doesn't contain the symbol `png_user_version_check` 
```
In [9]: p.loader.main_bin.get_symbol('png_user_version_check') is None
Out[9]: True
```

**Comment by frapik99** [comment]
@axt , thanks for you answer and sorry for my late response. I'm sorry that I was not clear but ```png_user_version_check``` is within the ```libpng16.so``` file. Can you try just loading the so and generate the graph?

Thanks for your time

**Comment by frapik99** [comment]
I managed to create the ```CFGAccurate``` with ```context_sensitivty_level=2``` but when I try to create the graph for ```png_user_version_check``` using ```plot_cfg``` I have the following error:

```
/root/.virtualenvs/angr/lib/python2.7/site-packages/bingraphvis/base.pyc in remove_node(self, node)
    132 
    133     def remove_node(self, node):
--> 134         self.nodes.remove(node)
    135         self.edges = filter(lambda edge: edge.src != node and edge.dst != node, self.edges)
    136 

KeyError: <bingraphvis.base.Node object at 0x7f4390363d10>

```

How can help to fix it?

Thanks

**Comment by axt** [comment]
Can you attach your version of libpng? My version doesn't contains the function `png_user_version_check`:

```
readelf -s libpng16.so.16  |grep png_user |wc -l
0
```

**Comment by frapik99** [comment]
Sure, thanks.
[libpng.zip](https://github.com/axt/angr-utils/files/1065788/libpng.zip)

**Comment by frapik99** [comment]
@axt , I managed to have the graph of the function and if the CFG is generated from the beginning of ```png_user_version_check``` all is fine but if I generate the whole ```CFGAccurate``` with ```context_sensitivity_level``` set to 2 I think does not work. I mean that there is an unexpected loop at the beginning of the function.

How can I generate the ```base.Graph``` using bingraphviz only for a specific function once I have the whole CFG? Could you generate the graph in both cases please? Maybe I'm doing something stupid.

Thanks

**Comment by frapik99** [comment]
Hi,
I have generated the ```CFGAccurate``` with ```context_sensitivity_level=2``` and generated the image using ```plot_cfg``` and the result was the attached image. Can you tell me why the function at ```0x9003da0``` is not shown in the graph? I tried with the example in ```angr-utils``` and the function ```verify``` was shown. Any idea why?

Anyway I have generated the graph in this way ```plot_cfg(cfg, "%s_%s_cfg" % (name, func.name), asminst=True, vexinst=False, func_addr={addr:True}, debug_info=False, remove_imports=True, remove_path_terminator=True)```

Thanks
![png_user_version_check1_cfg2](https://user-images.githubusercontent.com/22319973/27051119-25dd56d4-4fdf-11e7-8776-414ac0e3c141.png)

**Comment by axt** [comment]
Hi! Sorry for the delay, I was quite busy lately (and still I am). 
I will try to check it the latest tomorrow, but just a quick-tip: probably `remove_imports=True` causes this behavior.

When using bingraphvis, you either can specify a filter function, when creating the plot:
```
vis.process(cfg.graph, filter=lambda node: node.obj.function_address in [0x1234,0x3324]) 
```
or you can transfrom the graph, by adding the following to the pipeline:
```
vis.add_transformer(AngrFilterNodes(lambda node: node.obj.function_address in  [0x1234,0x3324]))
```
The later is the same as specifying `func_addr={0x1234:True, 0x3324:True}` to `plot_cfg`.

I hope this can help. Will get back to you with the details tomorrow!

**Comment by axt** [comment]
Hi! I've created the graph, using this snippet of code:
```python
    proj = angr.Project("./libpng16.so.16", load_options={'main_opts': {'custom_base_addr':0x0}, 'auto_load_libs':False})
    sym = proj.loader.main_bin.get_symbol("png_user_version_check")
    addr = sym.addr
    start_state = proj.factory.blank_state(addr=addr)
    start_state.stack_push(0x0)
    cfg = proj.analyses.CFGAccurate(fail_fast=True, starts=[addr], initial_state=start_state, context_sensitivity_level=2, keep_state=True, call_depth=100, normalize=True)
    plot_cfg(cfg, "cfg" , asminst=True, vexinst=False, debug_info=False, remove_imports=True, remove_path_terminator=True)
```
also works with this:
```
    proj = angr.Project("./pngpixel", load_options={'auto_load_libs':True})
    addr = 0x90044de
    start_state = proj.factory.blank_state(addr=addr)
    start_state.stack_push(0x0)
    cfg = proj.analyses.CFGAccurate(fail_fast=True, starts=[addr], initial_state=start_state, context_sensitivity_level=2, keep_state=True, call_depth=100, normalize=True)
    plot_cfg(cfg, "cfg" , asminst=True, vexinst=False, debug_info=False, remove_imports=True, remove_path_terminator=True)
```

It contains the function you are missing:
![cfg](https://user-images.githubusercontent.com/759365/27140201-43ecec38-5125-11e7-9917-9a9ce34467a1.png)

The line you posted that you used, adds a filter for the function selected by `addr`, which means it will remove every node from the graph which doesn't belong to that function, that is why you don't see the node `0x9003da0` on your plot, because it belongs to the '__x86.get_thunk.bx' function.

I'm a bit lost in this thread, can you summarize me, what kind of graphs do you want me to create? 
Or does this answer your problems?

**Comment by frapik99** [comment]
@axt , perfect exactly what I needed. One curiosity, why did you push 0 on the stack? What the advantage of having a normalized graph? In that function I can see the same number of blocks but more edges...

Also, Is it possible to have the graph of one function considering also all the blocks belonging to the all the functions called by the initial functions? I would like to have the graph ```png_user_version_check``` but considering the ```CFGAccurate``` generated on the whole binary.

Thanks

**Comment by frapik99** [comment]
Sorry, if you see the attached image the function at ```0x9003da0``` is not present. I have generated the ```CFGAccurate``` with ```context_sensitivity_level=2```. Do you think that the function was excluded because of it? Last question I promise :)

![cfg](https://user-images.githubusercontent.com/22319973/27197744-9ab3af4c-5239-11e7-9b5b-4d9d7c8f9d75.png)

**Comment by frapik99** [comment]
All sorted :) I used again func_addr

Thanks
